### PR TITLE
Limit superpmi replay failures using 'failureLimit' command line option

### PR DESCRIPTION
**Pull request by kunalspathak**
Limit the number of superpmi replay failures ~to 100~ to `failureLimit` command-line option after which we would skip compiling remaining methods.

If there is any obvious bug in the implementation such that every other method fails to compile, we might end up generating 1000s of `.mc` files and they further get copied from `%TEMP%` to spmi-collection folder. Because of this, the completion time of `superpmi replay` command is huge with lot of disk I/O. Instead, cap the failure to ~`100`~ `failureLimit`, in which case it is highly likely that there is a obvious bug that developer should fix before rerunning the replay command again.

Hopefully, it will be useful.

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
This works for replay, but not for asm diffs. In that case, we'll have a failed `res2`. We haven't worried about that one failing in the past, since if it fails, then either JIT1 will fail (and hit your code path), or JIT1 will succeed but we'll see an asm diff. I think you'll want to add some counting of JIT2 failures and a break-out, around:
```
            if (res2 == JitInstance::RESULT_ERROR)
            {
                LogError("JIT2 main method %d of size %d failed to load and compile correctly.",
                         reader->GetMethodContextIndex(), mc->methodSize);
            }
```

**Comment by kunalspathak** [comment]
I see. So if I understand correctly, I need to keep the existing check (for `replay` cases, the way you pointed) and add check around `res2 == JitInstance::RESULT_ERROR` to catch `asmdiffs` cases.

**Comment by BruceForstall** [comment]
Yes. In the asm diffs case, there are two exit codes. So for the JIT2/res2 case, you probably also need an `errorCount2`

**Comment by sandreenko** [comment]
Can this be an optional flag? I often use spmi replay when expecting 100+ failures but I want the full replay so I can choose the smallest mc file to debug.

**Comment by kunalspathak** [comment]
> Can this be an optional flag? I often use spmi replay when expecting 100+ failures but I want the full replay so I can choose the smallest mc file to debug.

That is a good feedback. Even I didn't like the hardcoded limit of 100, so we can have it more configurable.

File: src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp [hunk 1]
```diff
@@ -471,6 +475,11 @@ bool CommandLine::Parse(int argc, char* argv[], /* OUT */ Options* o)
                     }
                 }
             }
+            else if ((_strnicmp(&argv[i][1], "failureLimit", argLen) == 0))
+            {
+                ++i;
+                o->failureLimit = atoi(argv[i]);
```
Comment (BruceForstall) [inline src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp:481]: You need to check that someone actually provided an argument, e.g.,
```
                if (i >= argc)
                {
                    DumpHelp(argv[0]);
                    return false;
                }
```

File: src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp [hunk 2]
```diff
@@ -100,6 +100,10 @@ void CommandLine::DumpHelp(const char* program)
     printf("     If 'workerCount' is not specified, the number of workers used is\n");
     printf("     the number of processors on the machine.\n");
     printf("\n");
+    printf(" -failureLimit [limit]\n");
```
Comment (BruceForstall) [inline src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp:103]: The brackets imply that `limit` is optional. I would remove them, or replace them with `<` and `>` so:
```
printf(" -failureLimit <limit>\n");
```

File: src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp:481]: Seems like you should also ensure failureLimit >= 1

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/ToolBox/superpmi/superpmi/commandline.cpp:481]: ah, my bad.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**